File: solutions/go/src/playground/algorithm/longest_consecutive_sequence.go
package algorithm

import "math"

func longestConsecutive(nums []int) int {
	longest, m := 0, make(map[int]int)
	for _, n := range nums {
		if _, ok := m[n]; ok {
			continue
		}
		l, r := m[n-1], m[n+1]
		m[n] = l + r + 1
		m[n-l], m[n+r] = m[n], m[n]
		longest = int(math.Max(float64(longest), float64(m[n])))
	}
	return longest
}

File: solutions/java/src/main/java/playground/algorithm/LongestConsecutiveSequence.java
package playground.algorithm;

import java.util.HashMap;
import java.util.Map;

public class LongestConsecutiveSequence {

    public int longestConsecutive(int[] nums) {
        int longest = 0, left, right, sum;
        Map<Integer, Integer> map = new HashMap<>();
        for (int n : nums) {
            // 跳过已经出现过的数字
            if (map.containsKey(n)) {
                continue;
            }
            left = map.getOrDefault(n - 1, 0);
            right = map.getOrDefault(n + 1, 0);
            sum = left + right + 1;
            map.put(n, sum);
            // 更新序列头
            map.put(n - left, sum);
            // 更新序列尾
            map.put(n + right, sum);
            longest = Math.max(longest, sum);
        }
        return longest;
    }
}
